aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -66,11 +58,11 @@ static void copy_device_to_device(TensorIterator& iter, bool non_blocking) {
         cudaMemcpyDeviceToDevice,
         copy_stream));
   } else {
-    AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBool, iter.dtype(0), "copy_", [&] {
-      using dst_t = scalar_t;
-      AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBool, iter.dtype(1), "copy_", [&] {
-        copy_kernel_impl<dst_t, scalar_t>(iter);
-      });
+    // this is done intentionally done after build because copy has a "promotion"
+    // rule that always "promote" to target dtype.
+    iter.promote_common_dtype();
+    AT_DISPATCH_ALL_TYPES_AND3(kHalf, kBool, kBFloat16, iter.dtype(0), "copy_", [&] {
```

Comment:
I see a PR trying to enable BFloat16 for CUDA:
https://github.com/pytorch/pytorch/pull/27259/files#diff-6684cb81a1865b7d52d9f2f1789cd0ceR68-R70
So I include it in this PR also, so that I can benchmark